packages/deployer/src/compiler.ts

```diff
@@ -105,27 +115,48 @@ export class Compiler {
         await createDirIfDoesNotExistAsync(SOLC_BIN_DIR);
-        this._contractSources = await Compiler._getContractSourcesAsync(this._contractsDir);
+        this._contractSources = {};
+        const contractDirs = Array.from(this._contractDirs.values());
+        for (const contractDir of contractDirs) {
+            const sources = await Compiler._getContractSourcesAsync(contractDir.path, contractDir.path);
+            _.forIn(sources, (source, sourceFilePath) => {
+                const sourceFileId = constructUniqueSourceFileId(contractDir.namespace, sourceFilePath);
+
+                // Record the file's source and data
+                if (!_.isUndefined(this._contractSources[sourceFileId])) {
+                    throw new Error(`Found duplicate source files with ID '${sourceFileId}'`);
+                }
+                this._contractSources[sourceFileId] = source;
+
+                // Create a mapping between the contract id and its source file id
+                const contractId = constructContractId(contractDir.namespace, sourceFilePath);
+                if (!_.isUndefined(this._contractIds[contractId])) {
+                    throw new Error(`Found duplicate contract with ID '${contractId}'`);
+                }
+                this._contractIds[contractId] = sourceFileId;
+            });
+        }
+
         _.forIn(this._contractSources, this._setContractSpecificSourceData.bind(this));
-        const fileNames = this._specifiedContracts.has(ALL_CONTRACTS_IDENTIFIER)
-            ? _.keys(this._contractSources)
+        const contractIds = this._specifiedContracts.has(ALL_CONTRACTS_IDENTIFIER)
+            ? _.keys(this._contractIds)
             : Array.from(this._specifiedContracts.values());
-        for (const fileName of fileNames) {
-            await this._compileContractAsync(fileName);
+        for (const contractId of contractIds) {
+            await this._compileContractAsync(this._contractIds[contractId]);
         }
     }
     /**
      * Compiles contract and saves artifact to artifactsDir.
-     * @param fileName Name of contract with '.sol' extension.
+     * @param sourceFileId Unique ID of the source file.
      */
-    private async _compileContractAsync(fileName: string): Promise<void> {
+    private async _compileContractAsync(sourceFileId: string): Promise<void> {
         if (_.isUndefined(this._contractSources)) {
             throw new Error('Contract sources not yet initialized');
         }
-        const contractSpecificSourceData = this._contractSourceData[fileName];
-        const currentArtifactIfExists = await getContractArtifactIfExistsAsync(this._artifactsDir, fileName);
-        const sourceHash = `0x${contractSpecificSourceData.sourceHash.toString('hex')}`;
-        const sourceTreeHash = `0x${contractSpecificSourceData.sourceTreeHash.toString('hex')}`;
+        const contractSpecificSourceDataByFileId = this._contractSourceDataByFileId[sourceFileId];
```

Comment:
check for existence here also rename to `contractSpecificSourceDataIfExists` since it may be of type `ContractSourceData` or `undefined`